### DISCUSSION TITLE
Building a Payment Streaming Decentralized App with Soroban

### DISCUSSION THREAD

**Discussion by tosinshada** [stellar/sorobounty-spectacular#9] [General]
Hi,

I would like to share the first part of a series of decentralized Applications built on Soroban. Smart contracts have allowed us to make payments in new and unique ways one of which is the ability to stream payments to a recipient over some time.
- **Codebase:** The codebase is available [here](https://github.com/TosinShada/tide-soroban-contract) and hosts both the smart contract and the frontend
- **Tutorial:** The tutorial can be found [here](https://dev.to/shada/soroban-series-1-building-a-payment-streaming-app-156b) and it details some of the key functions in the smart contract.
- **Video Walkthrough:** I also did a video walkthrough of both the code base and the application which can be found [here](https://www.youtube.com/watch?v=wYqhZxqgx94)
- **Demo:** The current version of the application is hosted [here](https://tide-soroban-contract-frontend.vercel.app)

Looking forward to building other decentralized applications in this series.

**Comment by briwylde08** [comment]
Thank you so much for your submission. This looks great! We will review and get back to you within a week.

**Comment by kalepail** [comment]
When trying to create a stream I'm getting this error
<img width="411" alt="Screenshot 2023-08-28 at 5 47 04 PM" src="https://github.com/stellar/sorobounty-spectacular/assets/4383610/b9b19108-4f07-4efa-8408-3a930d7f41fa">

**Reply by tosinshada** [reply]
Hi, this might be because you don't have the token meant for streaming. Kindly share the address you are using to stream so that I can send you some

**Reply by kalepail** [reply]
`GAN352WSXTCDDTQEMJEVLT6B4LMUIGWIPBINRI3LUHAATKD72UAQYL5W`

Seems like a bug though which would make it really tricky for folks to test this dapp. There should be an easy method to go through using this without needing to have you manually send an asset.

Is there a currently documented way to do this that I'm just currently missing?

**Reply by tosinshada** [reply]
I totally agree especially since the errors being passed from the smart contract are not yet showing on the demo application properly. I have added a mint token feature on the page which should make the process a bit more seamless as well as some form validations.
You can directly mint some tokens and test the streaming feature.
Looking into the future, there might need to be a feature to wrap tokens and allow participants to stream the wrapped token similar to superfluid.

**Reply by kalepail** [reply]
Still getting the same error :/

**Reply by tosinshada** [reply]
Try again and mint some tokens before streaming. Vercel just published my changes some minutes ago.
Also stay within 5 days streaming period as the balance will be checked on the smart contract.
![Screenshot 2023-08-29 at 7 30 44 PM](https://github.com/stellar/sorobounty-spectacular/assets/89518803/1813d4f6-68e5-4a53-8226-8602f1e4fc96)

**Reply by kalepail** [reply]
I have. Same error
<img width="618" alt="Screenshot 2023-08-29 at 2 38 14 PM" src="https://github.com/stellar/sorobounty-spectacular/assets/4383610/73190064-5c6a-4335-9ae2-0d3de53546a0">

**Reply by tosinshada** [reply]
Also, note that self-streaming is not allowed. So the recipient cannot be the sender wallet signing the transaction. I will push a frontend validation for that shortly. Please confirm if you are streaming to a different address.

**Reply by kalepail** [reply]
Was definitely trying to self pay. Just tried using a different address. Same error again :/

**Reply by kalepail** [reply]
This was the values that error'd
<img width="623" alt="Screenshot 2023-08-29 at 2 44 32 PM" src="https://github.com/stellar/sorobounty-spectacular/assets/4383610/a1c9ac08-5c5e-409c-a082-450d1b831193">

**Reply by kalepail** [reply]
Might be date based? If I keep the date range to 3 days I can get a Flow Rate/Second to 2

**Reply by tosinshada** [reply]
Yes, the mint button will supply 1 million tokens to your wallet which should cover you for approximately 1 week of streaming 1 token per second

**Reply by kalepail** [reply]
Can you enter a float in the payments per second? Or only whole numbers 1 or 2?

**Reply by tosinshada** [reply]
You can use a float in the payments per second

**Reply by kalepail** [reply]
Alright I think I'm finally starting to get it. I would strongly encourage putting some work into the dapp to explain this all a bit more. (add a float point placeholder, add some text walking through how to configure the values, etc. Right now most folks are going to get stuck and frustrated.)

**Reply by tosinshada** [reply]
Thanks a lot for the suggestions, I have added validation for self-transfer and some messages about possible durations. I will add the floating point placeholder now.

**Comment by briwylde08** [comment]
Hi @TosinShada! We have reviewed your submission. [🎉](https://emojipedia.org/party-popper/)

Given our guidelines, your submission may qualify for an award of $1,000 worth of XLM. To receive that award, please fill out [this form](https://stellarform.typeform.com/to/HBjEyCJt) completely and accurately.

Thanks again! Please let us know if you have any questions.

Best,
Bri

**Comment by tosinshada** [comment]
Hi Bri, thanks for the award, I have filled out the form. 

I wanted to find out if there's any part of the guidelines I might have missed in this submission.

**Reply by briwylde08** [reply]
Hi again! I'm so glad you asked because we missed the demo video. After reviewing again, your submission qualifies for the full $1,500 worth of XLM.

Thanks for also providing the requested information in the form. You’ve passed KYC and I will log your award payment of $1,500 worth of XLM, which will be sent to the wallet you designated.

Thanks again and please let me know if you have any questions!

Best,
Bri

**Reply by tosinshada** [reply]
Woah thanks a lot. Looking forward to building  more on Soroban
